packages/seed-bible/db/annotations/library.tsx

```diff
@@ -268,3 +268,406 @@ export async function loadTranslationDocument(recordName: string, bookId: string
+ * @param endTime The end time in unix seconds to filter the reading history events.
+ */
+async function getYearlyReadingHistorySummary(recordName: string, year: number, startTime: number, endTime: number): Promise<any> {
+    let summary: ReadingHistorySummary = {
+        uniqueBooksRead: 0,
+        uniqueChaptersRead: 0,
+        totalTimeSpentReading: 0,
+        users: {}
+    };
+
+    const doc = await getReadingHistoryDocument(recordName, year);
+    const eventsArray = doc.getArray('events').type;
+    
+    for(let i = 0; i < eventsArray.length; i++) {
+        const e: SharedMap<any> = eventsArray.get(i);
+        const event: ReadingEvent = {
+            userId: e.get('userId'),
+            bookId: e.get('bookId'),
+            chapter: e.get('chapter'),
+            start: e.get('start'),
+            end: e.get('end'),
+        };
+
+        // Skip events outside the specified time range
+        // Only consider start time for filtering to avoid skipping time sections
+        if (event.start < startTime || event.start > endTime) {
+            continue;
+        }
+
+        const length = event.end - event.start;
+        summary.totalTimeSpentReading += length;
+        const userSummary = (summary.users[event.userId] ??= {
+            uniqueBooksRead: 0,
+            uniqueChaptersRead: 0,
+            totalTimeSpentReading: 0,
+            books: {}
+        });
+
+        userSummary.totalTimeSpentReading += length;
+        const bookSummary = (userSummary.books[event.bookId] ??= {
+            uniqueChaptersRead: 0,
+            totalTimeSpentReading: 0,
+            chapters: {}
+        });
+        
+        bookSummary.totalTimeSpentReading += length;
+
+        const chapterEvents = (bookSummary.chapters[event.chapter] ??= []);
+        chapterEvents.push(event);
+    }
+
+    updateSummaryTotals(summary);
+
+    return summary;
+}
+
+function updateSummaryTotals(summary: ReadingHistorySummary) {
+    // After processing all events, calculate uniqueChaptersRead
+    for (const userId in summary.users) {
+        const user = summary.users[userId];
+        for (const bookId in user.books) {
+            const book = user.books[bookId];
+            user.uniqueBooksRead += 1;
+            user.uniqueChaptersRead += Object.keys(book.chapters).length;
+            summary.uniqueBooksRead += 1;
+            summary.uniqueChaptersRead += Object.keys(book.chapters).length;
+        }
+    }
+}
+
+/**
+ * Finds the most recent reading event for the given user, book, and chapter.
+ * @param events The list of reading events.
+ * @param userId The ID of the user.
+ * @param bookId The ID of the book.
+ * @param chapter The chapter number.
+ * @returns The most recent reading event, or null if no event was found.
+ */
+function findMostRecentReadingEvent(events: SharedArray<SharedMap<any>>, userId: string, bookId: string, chapter: number, oldestTime: number): ReadingEvent | null {
```

Comment:
Maybe we might want to put the oldestTime comprobation above the userId/bookId/chapter comprobation. If not, the function will always return an event that matches the criteria, even if it is outside of the specified threshold.